float64 h
float64 hdot